### DISCUSSION TITLE
Issues compiling OpenCascade as wasm with emscripten

### DISCUSSION THREAD

**Discussion by wsbd008** [Open-Cascade-SAS/OCCT#40] [Q&A] [answered]
Hi,

Glad to see OCC choose Github as a new option for collatoration. Many thanks to the community!

I'm compiling latest OpenCascade (7.8.1) with latest emscripten (3.1.64), and get the following error:
```
[100%] Linking CXX executable ..\..\lin32\clang\bin\DRAWEXE.js
error: /tcl8.6@/tcl8.6.4/library does not exist
em++: error: 'D:\emsdk\upstream\emscripten\tools\file_packager.bat ..\..\lin32\clang\bin\DRAWEXE.data --from-emcc --preload D:/dev/OCCT/src/DRAWEXE/../DrawResources@/DrawResources /tcl8.6@/tcl8.6.4/library' failed (returned 1)
mingw32-make.exe[2]: *** [src\DRAWEXE\CMakeFiles\DRAWEXE.dir\build.make:161: lin32/clang/bin/DRAWEXE.js] Error 1
mingw32-make.exe[1]: *** [CMakeFiles\Makefile2:2943: src/DRAWEXE/CMakeFiles/DRAWEXE.dir/all] Error 2
mingw32-make.exe: *** [Makefile:135: all] Error 2
```
I was able to compile older version of OCC(7.6.0) with older version of emscripten(2.0.*). But after upgrade, I get this error.

Is there any way to get rid of this, or any options I can add to emcmake/emmake to try?

Thanks again.

**Comment by oursland** [comment]
I do not have an answer for you, but have you seen [opencascade.js](https://github.com/donalffons/opencascade.js)?  The build system for it is also [emscripten in a Docker container and pinned to 7.6.2](https://github.com/donalffons/opencascade.js/blob/5ff2b750ba4b9a9fdfbff8842712cbb562e78ce7/Dockerfile#L40C27-L40C67).  There may be other users of that project that could assist.

**Comment by wsbd008** [comment]
Thanks a lot for your reply. Yes I'm aware of opencascade.js. It is target for JavaScript programmer to build modeling apps directly on web.

My case is that to build the whole system with C++, and compile all dependencies (including OpenCascade) to wasm then link to the final html/wasm . So raising this issue.

**Reply by dpasukhi** [reply]
Hello @wsbd008, mostly our Wasm builds avoid DRAW, but it works. Probably the issue inside 3rd-party or CMake.
The same approach (avoiding DRAW) used current implementation for vcpkg (it is a easiest tool to build everything in one place). Because it is difficult to manage TCL+TK on wasm.
Could you please share how do you compile TK for wasm?
TCL is not required for build.

**Reply by wsbd008** [reply]
Hi @dpasukhi thanks a lot for you reply. I'm not compiling TK, excluded by CMake flag '-DUSE_TK=OFF'. 
How can I avoid DRAW? Seeing some patching process for DRAWEXE but didn't see a option to turn it off in the CMakeLists.txt.

**Reply by dpasukhi** [reply, marked as answer]
@wsbd008 my mistake. I mixed up tcl and tk. TCL is required for DRAW. TK is optional.
CMake Flag for disable DRAW(console APP)
`-DBUILD_MODULE_Draw=OFF`
The same workaround done for VCPKG.

**Reply by wsbd008** [reply]
I'm able to build OCC as wasm now, with the option '-DBUILD_MODULE_Draw=OFF'.
Thanks a lot @dpasukhi !
